src/ray/gcs/client_test.cc

```diff
@@ -233,191 +278,413 @@ TEST_F(TestGcsWithAsio, TestTaskTable) {
+      auto data = std::make_shared<protocol::TaskT>();
+      data->task_specification = task_spec;
+      RAY_CHECK_OK(client->raylet_task_table().Add(job_id, task_id, data, nullptr));
+    }
+    // Request notifications again. We should receive a notification for the
+    // current value at the key.
+    RAY_CHECK_OK(client->raylet_task_table().RequestNotifications(
+        job_id, task_id, client->client_table().GetLocalClientId()));
+  };
 
+  // Subscribe to notifications for this client. This allows us to request and
+  // receive notifications for specific keys.
+  RAY_CHECK_OK(client->raylet_task_table().Subscribe(
+      job_id, client->client_table().GetLocalClientId(), notification_callback,
+      subscribe_callback));
   // Run the event loop. The loop will only stop if the registered subscription
-  // callback is called for both writes to the object key.
+  // callback is called for the requested key.
   test->Start();
-  // Check that we received one callback for subscription success and two
-  // callbacks for the Add notifications.
-  ASSERT_EQ(test->NumCallbacks(), 3);
+  // Check that we received a notification callback for the first and least
+  // writes to the key, since notifications are canceled in between.
+  ASSERT_EQ(test->NumCallbacks(), 2);
 }
 
-TEST_F(TestGcsWithAe, TestSubscribeId) {
+TEST_F(TestGcsWithAe, TestTableSubscribeCancel) {
   test = this;
-  TestSubscribeId(job_id_, client_);
+  TestTableSubscribeCancel(job_id_, client_);
 }
 
-TEST_F(TestGcsWithAsio, TestSubscribeId) {
+TEST_F(TestGcsWithAsio, TestTableSubscribeCancel) {
   test = this;
-  TestSubscribeId(job_id_, client_);
+  TestTableSubscribeCancel(job_id_, client_);
 }
 
-void TestSubscribeCancel(const JobID &job_id,
-                         std::shared_ptr<gcs::AsyncGcsClient> client) {
-  // Write the object table once.
+void TestLogSubscribeCancel(const JobID &job_id,
+                            std::shared_ptr<gcs::AsyncGcsClient> client) {
+  // Add a log entry.
   ObjectID object_id = ObjectID::from_random();
+  std::vector<std::string> managers = {"jkl", "mno", "pqr"};
   auto data = std::make_shared<ObjectTableDataT>();
-  data->managers.push_back("A");
-  RAY_CHECK_OK(client->object_table().Add(job_id, object_id, data, nullptr));
+  data->manager = managers[0];
+  RAY_CHECK_OK(client->object_table().Append(job_id, object_id, data, nullptr));
 
-  // The callback for subscription success. Once we've subscribed, request
-  // notifications for the second object that was added.
-  auto subscribe_callback = [job_id, object_id](gcs::AsyncGcsClient *client) {
-    test->IncrementNumCallbacks();
-    // Request notifications for the object. We should receive a notification
-    // for the current value at the key.
+  // The callback for a notification from the object table. This should only be
+  // received for the object that we requested notifications for.
+  auto notification_callback = [object_id, managers](
```

Comment:
note that this is copying managers; for the test that's fine of course, but in rest of the codebase we wouldn't want to do that (instead, passing a shared pointer)